provider/inspec.rb

```diff
@@ -76,25 +76,25 @@ def generate_properties(data, props)
     def generate_property_files(prop_map, data)
       prop_map.flatten.compact.each do |prop|
         compile_file_list(
-          data[:output_folder],
+          data.output_folder,
           { prop[:target] => prop[:source] },
-          prop
+          property: prop[:property]
         )
       end
     end
 
     # Generates InSpec markdown documents for the resource
     def generate_documentation(data, base_name, plural)
-      docs_folder = File.join(data[:output_folder], 'docs', 'resources')
+      docs_folder = File.join(data.output_folder, 'docs', 'resources')
 
       name = plural ? base_name.pluralize : base_name
-      generate_resource_file data.clone.merge(
-        name: name,
-        plural: plural,
-        doc_generation: true,
-        default_template: 'templates/inspec/doc_template.md.erb',
-        out_file: File.join(docs_folder, "google_#{data[:product].api_name}_#{name}.md")
-      )
+      data_new = data.clone
```

Comment:
Why do we use data_new sometimes and data others?

Reply:
This is where a refactor could be good.

The idea should be that we have one `FileTemplate` per resource and that `FileTemplate` just compiles a bunch of templates with the same underlying object.

We don't fully follow that pattern. We sometimes alter the `data` object and don't want those changes to exist when we continue with generation. That's why we have `data.clone` all over the place - we aren't aware of when we're changing the data object, so we just alter the copies.

I've added a comment in the proper places to explain this better.

Reply:
Having to split between modifying and cloning feels like the wrong abstraction vs. one where cloning everywhere is appropriate. Maybe `FileTemplate` being a template for a single file isn't the right abstraction seeing that 2/3 providers are generating more than 1 file per `FileTemplate`?

Reply:
I think this ends up being a name issue.

I'm thinking that `FileTemplate` isn't 1-1 with templates, but 1-1 with resources.

(Yes, this abstraction doesn't completely hold because, say, operations aren't a resource. Hence, why you have to clone + alter)

Reply:
Ehh, I guess it just feels thorny to [do nothing in some places](https://github.com/GoogleCloudPlatform/magic-modules/blob/c6e029c533e69232e4878f2489d19ebcfb255fc5/provider/terraform.rb#L119-L124), [modifying in some more](https://github.com/GoogleCloudPlatform/magic-modules/blob/c6e029c533e69232e4878f2489d19ebcfb255fc5/provider/terraform.rb#L151-L162), and [cloning in the rest](https://github.com/GoogleCloudPlatform/magic-modules/blob/c6e029c533e69232e4878f2489d19ebcfb255fc5/provider/terraform.rb#L171-L183). I won't block on this, but do you mind documenting on the object why we might use each pattern?

Reply:
I completely agree. Any ideas on how to do this better?

The only questions I was trying to answer with this was: "what variables do my templates have access to and how do I easily find out what those variables are?"

Reply:
Continuing to always clone is probably better for consistency's sake imo? I don't think that not changing that part of the implementation detracts from your goal.

Reply:
I (weakly) disagree just because we lose the sense of which generation steps move 1-1 with resources and which are 1-1 with products

Reply:
Not sure I follow- https://github.com/GoogleCloudPlatform/magic-modules/pull/1605/files#diff-574e58f7dece6f9f7973f0724e4a3ccbR95 and the other InSpec `copied_data` usages look 1:1 with resources to me.

Reply:
Not sure what I was trying to go at with that comment.

Fix made!